### ISSUE TITLE
Invalid CVocConf leads to 500 errors

### ISSUE THREAD

**Issue by transfluxus** [IQSS/dataverse#10631]
**What steps does it take to reproduce the issue?**

Change the CVocConf setting with

curl -X PUT --upload-file cvoc-conf.json http://localhost:8080/api/admin/settings/:CVocConf

* When does this issue occur?

I discovered two types of problems:

1.
When an item in the list of the cvoc-conf.json file contains a retrieval-uri that is not in 
`https://gdcc.github.io/dataverse-external-vocab-support/`
or a file that does not exist (in that github page) 
dataverse will crash with a 500 `Internal Server Error - An unexpected error was encountered, no more information is available. `
when the user tries to create a new dataset in a dataverse which uses that related metadatablock.
 
2.
If an item in that json is invalid (I guess it checks against the CVocConf.schema.json) the server stops working entirely redirecting to 
500 - Internal Server Error - An unexpected error was encountered, no more information is available. 

e.g. I removed the key: `js-url`

* Which page(s) does it occurs on?

1.

when going from

https://example.net/dataverse/ex_datav

to 
https://example.net/dataset.xhtml?ownerId={id}

2.

dataverse main/index page

* What happens?

both cases
500 errors

* To whom does it occur (all users, curators, superusers)?

* What did you expect to happen?

For both cases: The External Vocabulary script not running

**Which version of Dataverse are you using?**

v. 5.12.1 build 1122-cf90431
  

**Any related open or closed issues to this bug report?**

**Screenshots:**

No matter the issue, screenshots are always welcome.

To add a screenshot, please use one of the following formats and/or methods described here:

* https://help.github.com/en/articles/file-attachments-on-issues-and-pull-requests
*

**Comment by pdurbin** [comment]
@transfluxus thanks for the bug report. Can you please attach your server.log file? It should show more details about the 500 error.

**Comment by transfluxus** [comment]
Hi @pdurbin 

would it be in `/usr/local/payara5/glassfish/domains/domain1/logs`

there are a lot of logs, ranging back to 2022 and one named `server.log_2024-06-13T07-29-51`

attached.
[logs1.txt](https://github.com/user-attachments/files/15838273/logs1.txt)



understanding the schema and cvoc mechanism a bit better, that problem does not occur always. But for example this array item would trigger it  ("js-urlX" instead of "js-url" and "retrieval-filtering" missing)
```
   {
    "field-name": "extLCProject",
    "term-uri-field": "extLCProject",
    "js-urlX": "https://diskordier.netX/lcProject.js",
    "protocol": "localcontexts",
    "retrieval-uri": "https://localcontextshub.org/api/v1/projects/?search={0}",
    "allow-free-text": true,
    "prefix": "https://localcontextshub.org/",
    "managed-fields": {},
    "languages": "",
    "vocabs": {
      "localcontexts": {
        "uriSpace": "https://localcontextshub.org/"
      }
    }
  }
```

**Comment by qqmyers** [comment]
It would be a good addition to improve handling of this error in Dataverse. As a work-around, there is a schema at https://github.com/gdcc/dataverse-external-vocab-support/blob/main/examples/config/CVocConf.schema.json that can be used to validate a CVoc config file.

**Comment by transfluxus** [comment]
Yeah, the datetime of the log-file is also a bit of. I will clean that folder and try to cause the error again and see if there will be a new file.

Indeed, the schema file is very important, since it seems to be the only place, where all fields for a CVoc protocol are documented in detail.  I ended up using [https://www.jsonschemavalidator.net/](https://www.jsonschemavalidator.net/), which works well for quick validation against some schema.

I can make some notes into the guidelines and make a pull request.

**Comment by qqmyers** [comment]
That would be great. Part of features being 'experimental' is that we don't have as much documentation and perhaps limited error handling, so contributions to either are appreciated (here and/or in the external vocab repo).

**Comment by sbarbosadataverse** [comment]
Waiting on user to make a pull request

**Comment by sbarbosadataverse** [comment]
@transfluxus Do you have an update on this?

**Comment by transfluxus** [comment]
Regarding the documentation,
I think it makes sense to copy some parts of:
https://github.com/gdcc/dataverse-external-vocab-support/tree/main/docs
into the readme of the repo or maybe even into the main doc. and 
I hope I can get to it, after our meeting.


Regarding the implementation, I found this todo in the code:
https://github.com/IQSS/dataverse/blob/dddcf29188a5c35174f3c94ffc1c4cb1d7fc0552/src/main/java/edu/harvard/iq/dataverse/DatasetFieldServiceBean.java#L281

I guess it would be sufficient to validate each `JsonObject jo` against the object schema, not the whole array, to filter out individual, invalid cvoc specifications, but that would require adding another validator library as dependency to the project, since I haven't identified any yet in the pom file.

**Comment by cmbz** [comment]
2024/07/10
- What are next steps, folks? @transfluxus @pdurbin @sbarbosadataverse

**Comment by pdurbin** [comment]
@cmbz I believe the next step is for @transfluxus to make a pull request (anything you come up with is fine, really!) I don't think we need to heavily track this effort. We can probably remove it from the main board (34) and just look at the pull request whenever it comes in. No rush, that is. @cmbz @sbarbosadataverse what do you think?

**Comment by cmbz** [comment]
@pdurbin sounds good! @transfluxus Please give us a heads up when you create the pull request. Thanks!
